python/rainbow/simulators/proximal_contact/blocking.py

```diff
@@ -0,0 +1,142 @@
+import numpy as np
+import numba as nb
+import networkx as nx
+from collections import defaultdict
+
+
+@nb.njit(parallel=True, nogil=True, cache=True)
+def _get_edges(J):
+    """ Get the edges of the contact graph.
+
+    Args:
+        J (ArrayLike): The contact jacobi matrix.
+
+    Returns:
+        (ArrayLike, ArrayLike): The source and target of the edges.
+    """
+    K = J.shape[0] // 4 # The number of contact points, here the each contact point has 4 rows of J.
+    step_size = 24 # Here the step_size is 24 , because each contact point has 8 non-zero blocks and each block has 3 columns.
```

Comment:
I still do not understand why this is 24.

For rigid bodies, you have 6 DOFs per body, so a row in the J would have at most 12 non-zero values.

For soft bodies, you have at most 4 nodes contributing to a row, and a node has at most 3 DOFs so that too should give 12 non-zeros per row.